aries/aries_vcx_core/src/wallet/askar/askar_did_wallet.rs

```diff
@@ -0,0 +1,178 @@
+            self.insert_did(&mut tx, did, TMP_DID_CATEGORY, &verkey, None)
+                .await?;
+            tx.commit().await?;
+            key_from_base58(&new_key_name)
+        } else {
+            Err(AriesVcxCoreError::from_msg(
+                AriesVcxCoreErrorKind::WalletRecordNotFound,
+                format!("did not found in wallet: {}", did),
+            ))
+        }
+    }
+
+    async fn replace_did_key_apply(&self, did: &str) -> VcxCoreResult<()> {
+        let mut tx = self.backend.transaction(self.profile.clone()).await?;
+        if let Some(did_data) = self.find_did(&mut tx, did, TMP_DID_CATEGORY).await? {
+            let verkey_did = did_data.did_from_verkey();
+            tx.remove(TMP_DID_CATEGORY, did).await?;
+            tx.remove_key(&verkey_did).await?;
+            self.update_did(&mut tx, did, DID_CATEGORY, &verkey_did, None)
+                .await?;
+            tx.commit().await?;
+            return Ok(());
+        } else {
+            return Err(AriesVcxCoreError::from_msg(
+                AriesVcxCoreErrorKind::WalletRecordNotFound,
+                "temporary did key not found in wallet",
+            ));
+        }
+    }
+
+    async fn sign(&self, key: &Key, msg: &[u8]) -> VcxCoreResult<Vec<u8>> {
+        if let Some(key) = self
+            .backend
+            .session(self.profile.clone())
+            .await?
+            .fetch_key(&did_from_key(key.to_owned()), false)
+            .await?
+        {
+            let local_key = key.load_local_key()?;
+            let key_alg = SigType::try_from_key_alg(local_key.algorithm())?;
+            Ok(local_key.sign_message(msg, Some(key_alg.into()))?)
+        } else {
+            Err(AriesVcxCoreError::from_msg(
+                AriesVcxCoreErrorKind::WalletError,
+                "key not found",
+            ))
+        }
+    }
+
+    async fn verify(&self, key: &Key, msg: &[u8], signature: &[u8]) -> VcxCoreResult<bool> {
+        if let Some(key) = self
+            .backend
+            .session(self.profile.clone())
+            .await?
+            .fetch_key(&did_from_key(key.to_owned()), false)
+            .await?
+        {
+            let local_key = key.load_local_key()?;
+            let key_alg = SigType::try_from_key_alg(local_key.algorithm())?;
+            Ok(local_key.verify_signature(msg, signature, Some(key_alg.into()))?)
+        } else {
+            Ok(false)
+        }
+    }
+
+    async fn pack_message(
+        &self,
+        sender_vk: Option<Key>,
+        recipient_keys: Vec<Key>,
+        msg: &[u8],
+    ) -> VcxCoreResult<Vec<u8>> {
+        if recipient_keys.is_empty() {
+            Err(AriesVcxCoreError::from_msg(
+                AriesVcxCoreErrorKind::InvalidInput,
+                "recipient keys should not be empty",
+            ))
+        } else {
+            let enc_key = LocalKey::generate(KeyAlg::Chacha20(Chacha20Types::C20P), true)?;
+            let packing = Packing::new();
```

Comment:
Changed, why do you think reusing a `Packing` struct might result in an undefined behaviour? Do you have a reproducer?

The only state that `Packing` has is a crypto box, which is itself stateless. If there is a problem with reusing the `SodiumCryptoBox` then it means the struct is probably not stateless (assuming the underlying crypto functions work correctly), which does not seem to be the case.

I am not sure if adding our specific functionality to a foreign type (`LocalKey`) in this case is in any measurable way better that having a dedicated struct we own, but that is just a matter of preference I guess.

Main reason for having the `Packing` struct was to have a single source of truth for the implementation of `crypto_box`, so swapping the implementation in the future would require only a single line change. It would also make it impossible to accidentally pack with one implementation and unpack with another. Now you need to hunt down all the occurrences of `SodiumCryptoBox::new` by search and replace in 2 files.